File: day06.py
"""Solution to day 6 for Advent of Code"""

from get_input import get_input


def part1(datastream, bufsize=4):
    buffer = []
    for c, char in enumerate(datastream, 1):
        buffer.append(char)
        if len(buffer) < bufsize:
            continue
        if len(buffer) > bufsize:
            buffer.pop(0)
        if len(set(buffer)) == bufsize:  # There's a duplicate somewhere
            return c
    raise NotImplementedError()


def part2(datastream):
    return part1(datastream, bufsize=14)


if __name__ == "__main__":
    assert part1("bvwbjplbgvbhsrlpgdmjqwftvncz") == 5
    assert part1("nppdvjthqldpwncqszvftbrmjlhg") == 6
    LINES = get_input(day=6, year=2022)
    print(f"Part 1: {part1(LINES)}")
    print(f"Part 2: {part2(LINES)}")

File: day06.rs
use common::get_input;
use std::collections::HashSet;

type Input = String;

fn find_start(buffer: &String, bufsize: usize) -> usize {
    let mut buf = Vec::new();
    for (i, item) in buffer.chars().enumerate() {
        buf.push(item);
        if buf.len() < bufsize { continue; }
        else if buf.len() > bufsize { buf.remove(0); }
        if buf.iter().collect::<HashSet<&char>>().len() == bufsize { return i + 1 }
    }
    unimplemented!()
}

fn part1(buffer: &Input) -> usize {
    find_start(buffer, 4)
}

fn part2(buffer: &Input) -> usize {
    find_start(buffer, 14)
}

fn main() {
    let input = get_input(06, 2022);
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
